content/api/servo-assembly/arduino-commands.md

```diff
@@ -0,0 +1,196 @@
+---
+title: Custom Arduino Commands
+---
+
+At the core of the servo assembly is an [Arduino Uno][arduino]. By default the
+Arduino runs firmware created by SourceBots providing simple access to the
+attached servo shield, compatible ultrasound sensors and to the GPIO pins
+directly.
+
+You can also extend the firmware on the Arduino to support more features, for
+example if you have other types of sensor which require higher precision timing
+than can be achieved through the Python API.
+
+## Requirements
+
+As your new firmware will still need to be able to talk to the Python code
+running on the Raspberry Pi, you'll need to base it on SourceBots default
+firmware. You can get a copy of the source code from GitHub either by
+downloading the [repo archive](servo-firmware-zip) or by cloning the repo:
+
+```txt
+git clone https://github.com/sourcebots/servo-firmware
+```
+
+We recommend using [Arduino Studio](https://www.arduino.cc/en/Main/Software) to
+edit your firmware as it provides direct support for building the code into the
+firmware binary and downloading it onto the Arduino.
+
+## Adding a command
+
+In order for the code running on your Pi to be able to interact with the
+functionality you're adding to the Arduino, you'll need to define a new command.
+
+Adding a command requires two changes. First you must define the function which
+will implement the operations of that command and then you register that command
+under the name you'll use to call it from the Python code.
+
+### Implementing the command
+
+Here's a simple command which just echoes back any arguments it receives:
+
+```cpp
```

Comment:
We should mention that this function definition must be placed before its use (the `commands` array), unless they add a prototype for the function as well of course.

Reply:
b93eaf0 reworks the start of the "Implementing the command" section and includes this.